clap_builder/src/output/help_template.rs

```diff
@@ -931,16 +942,83 @@ impl HelpTemplate<'_, '_> {
                 sub_help.write_flat_subcommands(subcommand, first);
             }
         }
+
+        // Commands under custom headings
+        if !custom_headings.is_empty() {
```

Comment:
This change is overall a lot more complicated than I would have expected and I'll likely need to dig into this a lot further to better understand how inherent the complexity is or isn't before we're able to move forward.

Reply:
That is a fair point.  If people rely on the help headings to logically group their commands, then consistently using that order elsewhere can make sense.  There can still be room for discussing whether it is worth it or not.  Most likely, we'll need to whittle away at other complexity first to see how much of an impact it really makes.

As for the other complexity, in each case where we are dealing with custom help headings, we are duplicating all of the underlying logic rather than folding the custom and built-in heading cases together, making this more ripe for bugs.  This code also special cases `help` which will make it diverge in order compared to anything else and seems like a potential feature independent of this work.

Reply:
> > There is another concern I had last night and that is whether there are any other artefacts that are caused by the feature.
> 
> What do you mean?

That there would be some other interaction from the help_heading API - I produced a more comprehensive set of tests (all the helps tests with subcommand with a help_heading added) and only found a couple of problems with the flattened versions. One fixed and the other I am still working on.